src/com/maddyhome/idea/vim/extension/textobjindent/VimIndentObject.java

```diff
@@ -0,0 +1,220 @@
+import static com.maddyhome.idea.vim.group.visual.VisualGroupKt.vimSetSelection;
+import static com.maddyhome.idea.vim.helper.StringHelper.parseKeys;
+
+/**
+ * Port of vim-indent-object:
+ * https://github.com/michaeljsmith/vim-indent-object
+ *
+ * <p>
+ * vim-indent-object provides these text objects based on the cursor line's indentation:
+ * <ul>
+ *   <li><code>ai</code> <b>A</b>n <b>I</b>ndentation level and line above.</li>
+ *   <li><code>ii</code> <b>I</b>nner <b>I</b>ndentation level (no line above).</li>
+ *   <li><code>aI</code> <b>A</b>n <b>I</b>ndentation level and lines above and below.</li>
+ *   <li><code>iI</code> <b>I</b>nner <b>I</b>ndentation level (no lines above and below). Synonym of <code>ii</code></li>
+ * </ul>
+ *
+ * See also the reference manual for more details at:
+ * https://github.com/michaeljsmith/vim-indent-object/blob/master/doc/indent-object.txt
+ *
+ * @author Shrikant Kandula (@sharat87)
+ */
+public class VimIndentObject implements VimExtension {
+
+  @Override
+  public @NotNull
+  String getName() {
+    return "textobj-indent";
+  }
+
+  @Override
+  public void init() {
+    putExtensionHandlerMapping(MappingMode.XO, parseKeys("<Plug>textobj-indent-ai"), getOwner(),
+      new IndentObject(true, false), false);
+    putExtensionHandlerMapping(MappingMode.XO, parseKeys("<Plug>textobj-indent-aI"), getOwner(),
+      new IndentObject(true, true), false);
+    putExtensionHandlerMapping(MappingMode.XO, parseKeys("<Plug>textobj-indent-ii"), getOwner(),
+      new IndentObject(false, false), false);
+
+    putKeyMapping(MappingMode.XO, parseKeys("ai"), getOwner(), parseKeys("<Plug>textobj-indent-ai"), true);
+    putKeyMapping(MappingMode.XO, parseKeys("aI"), getOwner(), parseKeys("<Plug>textobj-indent-aI"), true);
+    putKeyMapping(MappingMode.XO, parseKeys("ii"), getOwner(), parseKeys("<Plug>textobj-indent-ii"), true);
+  }
+
+  static class IndentObject implements VimExtensionHandler {
+    final boolean includeAbove;
+    final boolean includeBelow;
+
+    IndentObject(boolean includeAbove, boolean includeBelow) {
+      this.includeAbove = includeAbove;
+      this.includeBelow = includeBelow;
+    }
+
+    static class IndentObjectHandler extends TextObjectActionHandler {
+      final boolean includeAbove;
+      final boolean includeBelow;
+
+      IndentObjectHandler(boolean includeAbove, boolean includeBelow) {
+        this.includeAbove = includeAbove;
+        this.includeBelow = includeBelow;
+      }
+
+      @Override
+      public @Nullable
+      TextRange getRange(@NotNull Editor editor, @NotNull Caret caret, @NotNull DataContext context,
+                         int count, int rawCount, @Nullable Argument argument) {
+        final int caretLineNum = caret.getCaretModel().getVisualPosition().getLine();
+        String content = editor.getDocument().getText();
+        String[] lines = content.split("\n");
+        String caretLine = lines[caretLineNum];
```

Comment:
`caretLineNum` is a line taken from the visual position, but here you take the index for all lines. This may work incorrect if you'll have folded lines above your current line. I think there should be a `logicalPosition` for lines indexing (if you'll use this approach, see the comments above).

I think this post may explain the difference between the visual and logical positions in IJ: https://jetbrains.org/intellij/sdk/docs/tutorials/editor_basics/coordinates_system.html#caret-logical-position
In short, a logical position is a position in the text while a visual position is a position that is rendered for the user. If you have a 10 lines file and 3-6 lines are folded, this file has 10 logical lines, but 7 visual lines.